### PR TITLE
Adding setVariationAxis for variable fonts

### PR DESCRIPTION
**Pull request by themancalledjakob**
Hey,
first pull request here, I hope there is interest and I'm doing this right.

A very basic implementation of variable font axes.
slightly inspired by [godot](https://github.com/godotengine/godot/blob/master/modules/text_server_fb/text_server_fb.cpp) and [freetype docs](https://freetype.org/freetype2/docs/reference/ft2-multiple_masters.html).

There is a second pull request with a cli implementation for the main.cpp application in the making.

### CODE REVIEW

**Comment by wdanilo** [comment]
That's amazing! Supporting variable fonts would be incredibly useful. Is there plan for full variable fonts support sooner than latter @Chlumsky ? :)

**Comment by themancalledjakob** [comment]
hey @wdanilo in the meantime with my fork [(branch cli-variable-fonts)](https://github.com/themancalledjakob/msdfgen/tree/cli-variable-fonts) you can set arbitrary amount of axes for a font, and so far it worked fine with all opentype fonts that I tried. Of course, though I tried to put the code at a spot that hopefully makes sense, @Chlumsky might prefer a different implementation, so it's not future proof.

**Comment by wdanilo** [comment]
@themancalledjakob that's amazing. @Chlumsky, what do you think of the implementation on the @themancalledjakob branch? :)

**Comment by Chlumsky** [comment]
Hello, I was on vacation but I will try to look at it this week. However, I think I'm going to change at least the name of the function, and I think there should be a way to find out the list of names that can be used.

**Comment by themancalledjakob** [comment]
hey, thanks for the reply and hope you had a nice vacation.

> I think there should be a way to find out the list of names that can be used.

sure! There is also a way of not only putting out the list of names, but also the predefined masters with their corresponding axes coordinates.

It can be quite interesting, especially when a font has multiple axes and you want to know which versions are not interpolated.
e.g.:
```
master(1): 
    coords: Weight(100.000000) hght(100.000000) 
master(2): 
    coords: Weight(200.000000) hght(100.000000) 
master(3): 
    coords: Weight(300.000000) hght(100.000000) 
master(4): 
    coords: Weight(400.000000) hght(100.000000) 
master(5): 
    coords: Weight(500.000000) hght(100.000000) 
master(6): 
    coords: Weight(600.000000) hght(100.000000) 
master(7): 
    coords: Weight(700.000000) hght(100.000000) 
master(8): 
    coords: Weight(800.000000) hght(100.000000) 
master(9): 
    coords: Weight(900.000000) hght(100.000000) 
master(10): 
    coords: Weight(900.000000) hght(200.000000) 
master(11): 
    coords: Weight(900.000000) hght(300.000000) 
master(12): 
    coords: Weight(900.000000) hght(400.000000) 
master(13): 
    coords: Weight(900.000000) hght(500.000000) 
master(14): 
    coords: Weight(900.000000) hght(600.000000) 
master(15): 
    coords: Weight(900.000000) hght(700.000000) 
master(16): 
    coords: Weight(900.000000) hght(800.000000) 
master(17): 
    coords: Weight(900.000000) hght(900.000000) 
```

this might avoid some confusion, since in the above example setting Weight to 100 and hght to 800 won't have any effect. There are quite some fonts which have these kind of quirks.

should that be then an option like `-printmetrics` but then `-printvariations` (or similar)?

> However, I think I'm going to change at least the name of the function

Which name would you rather have for the function? I'd be happy to adjust it and send a new PR.
Do you prefer a single PR, or a PR per change?

**Comment by Chlumsky** [comment]
I meant that there should be just a function to fetch all the possible axis names and value ranges, not that there should also be a command line option for it. I will also have to update the old Windows build because it uses an old FreeType library that is committed in the repo and it probably doesn't support variable fonts.

Please just do one pull request, I will then merge it into a new branch, clean it up to my liking and then release it.

**Comment by themancalledjakob** [comment]
> I meant that there should be just a function to fetch all the possible axis names and value ranges, not that there should also be a command line option for it.

Got it. I will not touch the command line interface and just add the functions.

> I will also have to update the old Windows build because it uses an old FreeType library that is committed in the repo and it probably doesn't support variable fonts

Variable font support was added in freetype-2.8, that would be roughly 5 years ago. In both comitted win32 and win64 libs the necessary functions and a reference to freetype-2.9.1 are present. Don't have a windows machine here to verify it, but chances are good that you don't _need_ to update it. Not sure though if there are any bugfixes you'd want to pull in anyway.

> Please just do one pull request, I will then merge it into a new branch, clean it up to my liking and then release it.

Sounds great. I will be travelling today and have some time freed up for it during the weekend.

**Comment by themancalledjakob** [comment]
Hey, sorry for the delay.

There is now a simple function to fetch all the possible axis names and value ranges. I hesitated a bit to introduce a struct for the axis, but don't see another way to avoid dealing with freetype's number types.

You said you'd like to adjust it as you please, but let me know if there is something you'd like me to do.

File: ext/import-font.cpp [hunk 1]
```diff
@@ -215,4 +222,48 @@ bool getKerning(double &output, FontHandle *font, unicode_t unicode1, unicode_t
...
+}
+
+bool getVariationAxes(std::vector<FontVariationAxis> &axes, FontHandle *font, FreetypeHandle *library) {
+    if (font->face->face_flags & FT_FACE_FLAG_MULTIPLE_MASTERS) {
+        FT_MM_Var *amaster;
+        FT_Get_MM_Var(font->face, &amaster);
+
+        for (FT_UInt i = 0; i < amaster->num_axis; i++) {
+            FontVariationAxis axis{
+                amaster->axis[i].name,
```
Comment (Chlumsky) [inline ext/import-font.cpp:256]: Is it guaranteed that the pointer to name string remains valid after `FT_Done_MM_Var`?

**Comment by Chlumsky** [comment]
@themancalledjakob Please see the [feature-variable-fonts branch](https://github.com/Chlumsky/msdfgen/tree/feature-variable-fonts) and let me know if you have any comments. The CLI is used as `-varfont "myfont.ttf?var1=0.5&var2=1" 'A'`

I have two concerns though:
- The lifetime of `FontVariationAxis::name` (see above)
- and that the names seem to be different from e.g. the ones listed in https://v-fonts.com/ (`OpticalSize` vs `opsz` etc.).

**Pull request merged**